beluga/test/beluga/estimation/test_cluster_based_estimation.cpp

```diff
@@ -0,0 +1,373 @@
+  while (!prio_queue.empty()) {
+    const auto top = prio_queue.top();
+    EXPECT_GE(prev_weight, top.priority);
+    prev_weight = top.priority;
+    prio_queue.pop();
+  }
+}
+
+TEST_F(ClusterBasedEstimationDetailTesting, MapGridCellsToClustersStep) {
+  // create a map with four independent peaks
+  std::vector<std::tuple<double, double, double>> coordinates;
+  for (double x = 0.0; x < 36.0; x += 1.0) {
+    for (double y = 0.0; y < 36.0; y += 1.0) {
+      const auto weight = std::abs(std::sin(10.0 * x * M_PI / 180.0)) * std::abs(std::sin(10.0 * y * M_PI / 180.0));
+      coordinates.emplace_back(x, y, weight);
+    }
+  }
+
+  GridCellDataMap2D map;
+  for (const auto& [x, y, w] : coordinates) {
+    const auto state = SE2d{SO2d{0.}, Vector2d{x, y}};
+    map.emplace(spatial_hash_function(state), GridCellData{w, 0, state, std::nullopt});
+  }
+
+  const auto weight_cap = calculate_percentile_weight_threshold(map, 0.9);
+
+  cap_grid_cell_data_weights(map, weight_cap);
+
+  const auto adjacent_grid_cells = {
+      SE2d{SO2d{0.0}, Vector2d{+kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{-kSpatialHashResolution, 0.0}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, +kSpatialHashResolution}},
+      SE2d{SO2d{0.0}, Vector2d{0.0, -kSpatialHashResolution}},
+  };
+
+  // test proper
+
+  // only cells beyond the 10% weight percentile to avoid the messy border
+  // between clusters beneath that threshold
+  const auto ten_percent_threshold = calculate_percentile_weight_threshold(map, 0.15);
+  map_cells_to_clusters(map, spatial_hash_function, adjacent_grid_cells, weight_cap);
+
+  auto cells_above_minimum_threshold_view =
+      coordinates | ranges::views::filter([&](const auto& c) { return std::get<2>(c) >= ten_percent_threshold; });
+
+  auto quadrant_1_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) < 18.0 && std::get<1>(c) < 18.0;
+                         });
+  auto quadrant_2_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) >= 18.0 && std::get<1>(c) < 18.0;
+                         });
+  auto quadrant_3_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) < 18.0 && std::get<1>(c) >= 18.0;
+                         });
+  auto quadrant_4_view = cells_above_minimum_threshold_view | ranges::views::filter([&](const auto& c) {
+                           return std::get<0>(c) >= 18.0 && std::get<1>(c) >= 18.0;
+                         });
+
+  const auto coord_to_hash = [&](const auto& coords) {
+    const auto& [x, y, w] = coords;
+    const auto state = SE2d{SO2d{0.}, Vector2d{x, y}};
+    return spatial_hash_function(state);
+  };
+
+  const auto hash_to_id = [&](const auto& hash) { return map[hash].cluster_id ? map[hash].cluster_id.value() : 9999; };
```

Comment:
@glpuga hmm, shouldn't all particles be assigned a cluster? If so, why not simply invoke `map[hash].cluster_id.value()` unconditionally and let the exception propagate if broken?